packages/loader/container-loader/src/test/connectionStateHandler.spec.ts

```diff
@@ -1139,6 +1143,48 @@ describe("ConnectionStateHandler Tests", () => {
 		},
 	);
 
+	it("test 'read' reconnect & races ", async () => {
+		connectionStateHandler = createHandler(
+			false, // connectedRaisedWhenCaughtUp,
+			true, // readClientsWaitForJoinSignal
+		);
+
+		connectionStateHandler.establishingConnection({ text: "initial connect" });
+		connectionStateHandler.receivedConnectEvent(connectionDetails);
+		connectionStateHandler_receivedJoinSignalEvent(connectionDetails);
+		assert.strictEqual(
+			connectionStateHandler.connectionState,
+			ConnectionState.Connected,
+			"Client 1 should be in connected state",
+		);
+
+		connectionStateHandler.receivedDisconnectEvent({ text: "disconnect" });
+
+		connectionDetails2.mode = "read";
+		connectionStateHandler.establishingConnection({ text: "initial connect" });
+		connectionStateHandler.receivedConnectEvent(connectionDetails2);
+
+		// Clear Audience
+		connectionStateHandler_receivedLeaveSignalEvent(connectionDetails.clientId);
```

Comment:
Is this necessary for the test?

Reply:
Oh I see, it represents the removal of this from the audience. Of course.

If this line is removed, then connection 2 _does_ need to wait for connection 1 to leave, right?

Reply:
Ok I'm confused 😅  The redundant Join Signal (next line of code) adds the first clientId back to the Audience.  I'll go look at the code to see what's going on (started reviewing with tests first)

Reply:
I'm a little nervous about the logic there, given the volatility of leave/join signals.  It's quite a rigid state machine, and it may not fit anymore.

Rather than maintaining a state machine based on the sequence of the Join/Leave events (could be ops or signals), can we be quicker to check the audience at key points?

Specifically, the places that currently check if we're "waiting for the leave op" should first check if the old clientId is still in the quorum, and if not cancel the timer. I just don't trust that every client will get the full sequence of leaves and joins.

Definitely some FUD here, open to a reasonable explanation of what guarantees we do have with the signals.

Reply:
I'm simply simulating how ConnectionStateHandler observes reconnection flow. It will see Audience being fully cleared, and then repopulated with new state. I'm adding comment to clarify that.

Just to make sure we are on the same page: If client was connected as "read", we never wait for "leave" signal for it. So, number of times it shows up again in Audience (because we reconnect to different front-end) should not matter (that's what this PR fixes).

For "write" connections, we make determination (to raise timer and wait for it or not) only on disconnect and stick with this decision. We do not re-evaluate it, and if timer is active, we proceed to next step only when old clientId disappears from quorum.

I think we enforce what you are asking for - see applyForConnectedState():

```ts
		assert(
			!this.waitingForLeaveOp || this.hasMember(this.clientId),
			0x2e2 /* "Must only wait for leave message when clientId in quorum" */,
		);
```

and then there is only one place that cancels timer:

```ts
	private receivedRemoveMemberEvent(clientId: string) {
		// If the client which has left was us, then finish the timer.
		if (this.clientId === clientId) {
			this.prevClientLeftTimer.clear();
			this.applyForConnectedState("removeMemberEvent");
		}
	}
```

(Ideally it should check also if timer is present, I'll look more into it).

I actually do not see a reason to check at random places if client is still in quorum, as there should be no way for us to miss that event.

Let me know what you think